docs/design/features/greenthreads.md

```diff
@@ -0,0 +1,233 @@
+7. The existing Async/Await pattern
+  - The existing async/await pattern is both ubiquitous throughout the libraries and not entirely compatible with the green threading model. While it isn’t technically an implementation problem, it provides significant opportunity for customer confusion.
+  - We chose to integrate the api for these two constructs which ended up working out.
+8. Support for thread static variables
+  - Thread statics are commonly used within the BCL to achieve great performance. However, the green thread model in many languages assumes that each green thread has a unique set of thread statics.
+  - In .NET we have more usage of these variables, so we implemented a re-use model for thread statics. This mirrors how our ThreadPool gets the best performance.
+
+### Implementation Challenges
+During the implementation, we encountered numerous challenges.
+- The Windows OS is not designed to allow green threads. Making this work relies on somewhat undocumented behavior
+- Various diagnostic tools fail to work reliably (ETW, windbg, IDNA)
+- SEH exception handling does not safely handle cross stack segments
+- Diagnosing implementation issues has been very slow and difficult, as they tend to cause memory corruption across multiple threads
+- Every OS/Architecture pair needs quite a bit of custom hand-written assembly. While not yet known it is possible that MUSL based Linux may need different logic than glibc based Linux
+-	The prototype implementation was built on Windows due to team member familiarity with the platform, and therefore implementation challenges discovered were focused on the problems encountered in development on Windows. However, there are non-Windows issues that were discovered
+  - Thread local storage access on non-Windows platforms may utilize an unbounded amount of stack as it requires a call to a libc provided function in some cases
+  - Integration of Linux Perf with split stack is not supported as far as research has shown
+
+### Design of underlying runtime changes
+The underlying runtime was changed to support allocating stack segments and transitioning onto those segments when running on the thread pool. The implementation chosen was a fairly simple segmented stack approach, which is known to have various hot-split problems, but is generally considered the most straightforward way to implement a custom thread stack system.
+In the implementation, we chose to implement a unified stack with both local variables and the control stack together. This was chosen as it is somewhat simple to implement and did not violate any of the implementation constraints found. In order to support injected OS code onto the stack, as well as FCalls/HCalls, a Red Zone of about 4KB was added to each stack segment, bringing the minimum stack allocation to 12KB for a stack segment.
```

Comment:
"FCalls/HCalls" 

Would be good to point to a doc where these are defined/described.

Reply:
These are runtime-internal concepts, but there is a section in BOTR on them.

Reply:
Link has been added to the appropriate BOTR chapter